rfcs/sdk-spec.md

```diff
@@ -0,0 +1,1043 @@
+When a consumer function receives a batch of messages, it is responsible for processing each message in the batch and acknowledging that the message has been processed by deleting it from the queue.
+If a consumer function does not acknowledge a message, the message will be re-delivered to another consumer function.
+
+A queue is not FIFO (first-in-first-out), so messages may be delivered out of order.
+
+```ts
+struct QueueProps {
+  /**
+   * How long a message is available for consumption before it's made available
+   * for other consumers.
+   */
+  timeout: duration?;
+}
+
+struct QueueAddConsumerProps { /* elided */ }
+
+resource Queue {
+  init(props: QueueProps = {});
+
+  /**
+   * Run an inflight in a cloud function whenever a message is pushed to the queue. 
+   */
+  add_consumer(fn: inflight (message: Serializable) => void, opts: QueueAddConsumerProps?): void;
+
+  /**
+   * Push a message to the queue.
+   */
+  inflight push(message: Serializable): void;
+  /**
+   * Pops (deletes and returns) a message from the queue.
+   */
+  inflight pop(): Serializable;
+}
+```
+
+Future extensions:
+- automatic dead-letter queue configuration
+
+## Function
+
+The function resource represents a stateless function that can run a snippet of code whenever invoked (and any number of times in parallel).
+Functions are typically used to process data in response to events, such as a file being uploaded to a bucket, a message being pushed to a queue, or a timer expiring.
+
+When a function is invoked on a cloud provider, it is typically executed in a container that is spun up on demand.
+The container is then destroyed after the function finishes executing.
+This makes them easy to scale and fault-tolerant.
+
+Functions may be invoked more than once, and some cloud providers may automatically retry failed invocations.
+In addition, it is possible for functions to be partially invoked (e.g. if a container is destroyed mid-execution).
+For performance reasons, most cloud providers impose a timeout on functions, after which the function is automatically terminated.
+
+```ts
+struct FunctionProps {
+  /**
+   * The maximum amount of time the function can run, in seconds.
+   * @default 1min
+   */
+  timeout: duration?;
+
+  /**
+   * The amount of memory to allocate to the function, in MB.
+   * @default 128
+   */
+  memory: num?;
+
+  /**
+   * The maximum number of concurrent invocations of the function.
+   * @default 10
+   */
+  concurrency: num?;
+
+  /**
+   * The environment variables to pass to the function.
+   * @default {}
+   */
+  env: Map<str, str>?;
+}
+
+resource Function {
+  init(handler: inflight (input: any): any, props: FunctionProps = {});
+
+  /**
+   * Add an environment variable to the function.
+   */
+  add_env(key: str, value: str): void;
+
+  /**
+   * Invoke the function.
+   */
+  inflight invoke(payload: Serializable): Serializable;
```

Comment:
In my current understanding, it can return whatever the FaaS returns. For example, a function could do a quick calculation and return the result to whoever called `invoke`.

It's good you ask this because I realized, this only makes sense in cases where you are invoking the function **synchronously**. (e.g. someone calls `invoke` and awaits the function's response). In other cases, you might invoke a function and it might do some work taking several minutes -- so you just want to fire and forget. (In this case, a successful response just means that AWS Lambda received your invocation request -- but it doesn't contain the result).

I'll propose this in the spec as `invoke_async` -- let me know if this mental model makes sense, and if if you have any better name ideas 🙂

Reply:
Makes sense, thank you! Maybe you could add a `@returns` annotation to the function signature to explain this behavior to the API users?

Reply:
Updated - it should be a bit clearer now.